Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp

```diff
@@ -67,6 +67,20 @@ namespace AZ
             m_materialPipelineData = {};
             m_materialAsset = { &materialAsset, AZ::Data::AssetLoadBehavior::PreLoad };
             ShaderReloadNotificationBus::MultiHandler::BusDisconnect();
+            Data::AssetBus::Handler::BusDisconnect();
+            if (!m_materialAsset.IsReady())
+            {
+                // We will call this function again later when the asset is ready.
+                Data::AssetBus::Handler::BusConnect(m_materialAsset.GetId());
+                return RHI::ResultCode::Success;
+            }
```

Comment:
1. If `BusConnect()` for the material is only set here, and `Material::OnAssetReady()` already calls `BusDisconnect()`, isn't this `BusDisconnect()` redundant?
2. This logic may fit better in `CreateInternal()` - the only way we could get here is if you called `CreateInternal()` with a material asset that wasn't ready. The only other entry point is from shader asset notifications which aren't being listened to yet. Moving it would avoid this check in cases where it wasn't necessary.

Reply:
I thought about that too. BusDisconnect() is redundant. Will remove it... But the CreateInternal thing wont'work, as the rest of the code base call InstanceDatabase<Materia>>FindOrCreate() and assumes the Material is good to go. That's why Init() must check for IsReady(), if I don't, then it crashes because the callers, who created the material are not waiting for OnAssetReady... And there are many touch points I'd have to fix that call  InstanceDatabase<Materia>>FindOrCreate().

Reply:
So the creation function used for Material is just `CreateInternal()`, so all roads should lead there. (see `MaterialSystem::Init()`). The only other places I see `Init()` called is from the shader reloads, but those won't be listened for until the material has successfully initialized once. Is the behavior your describing something you anticipated would happen or something you actually saw happen? Not a blocking issue, but still seems like if you could do this in CreateInternal it'd be better.